Allow deletion of a previous values file key

scottrigby opened the pull request:
Since #1656 Helm deep-merges multiple values files. However, this also means there is currently no way to delete an existing key that may be incompatible with a newly added one.

This PR addresses option 2 in #1966. My Golang fu is not top notch yet, but that said let me know if this is wrong so I can fix it (PS Thanks to @rabellamy for sending me Go map documentation).

k8s-reviewable commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.kubernetes.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.kubernetes.io/reviews/kubernetes/helm/2648)
<!-- Reviewable:end -->

thomastaylor312 commented:
@scottrigby I think that you added the right logic in the wrong place. When I specify `foo: "null"`, I get this:
```
NAME:   awesome-panther
REVISION: 1
RELEASED: Wed Jul  5 21:30:57 2017
CHART: nginx-0.1.0
USER-SUPPLIED VALUES:
{}

COMPUTED VALUES:
foo: bar
image:
  pullPolicy: IfNotPresent
  repository: nginx
  tag: 1.11.0
index: <h1>Hello</h1> <p>This is a test</p>
nodeSelector: {}
podAnnotations: {}
replicaCount: 1
resources: {}
restartPolicy: Never
service:
  annotations: {}
  clusterIP: ""
  externalIPs: []
  loadBalancerIP: ""
  loadBalancerSourceRanges: []
  nodePort: ""
  port: 8888
  type: ClusterIP
sleepyTime: "10"
```

The function you modified only is merging user supplied values, not the `values.yaml` file. That merging is done on the tiller side (confusing I know) right [here](https://github.com/kubernetes/helm/blob/master/pkg/chartutil/values.go#L266). So I think if you move the logic to there it will work

seh commented on cmd/helm/install.go:
```diff
@@ -273,6 +273,14 @@ func (i *installCmd) run() error {
+		// If the new value is literally the string "null", remove the value's key.
+		// This allows helm's multiple value sources (value files or --set) to
+		// remove incompatible keys from the previous values.
+		if v == "null" {
```
- [ ] Note that "null" is only one of [the values that the YAML specification projects to null](http://www.yaml.org/spec/1.2/spec.html#id2803362).  
  There is also "Null," "NULL," and "~."

seh commented on cmd/helm/install.go:
```diff
@@ -273,6 +273,14 @@ func (i *installCmd) run() error {
 func mergeValues(dest map[string]interface{}, src map[string]interface{}) map[string]interface{} {
 	for k, v := range src {
+		// If the new value is literally the string "null", remove the value's key.
+		// This allows helm's multiple value sources (value files or --set) to
```
- [ ] s/helm's/Helm's/
- [ ] s/multiple value sources/various sources of values/

scottrigby commented:
@thomastaylor312 @seh I just updated this to cover earlier feedback.

Thanks @rabellamy for pairing on Golang syntax.

seh commented:
I'm trusting @thomastaylor312 that the accommodation is in the right place now.

---

Reviewed 2 of 2 files at r2.
Review status: all files reviewed at latest revision, 2 unresolved discussions.

---

*[pkg/chartutil/values.go, line 281 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2648#-KoO-684_YPaBDyFZJTm:-KoO-685emDbJlen5WVl:bn77y93) ([raw file](https://github.com/kubernetes/helm/blob/413a01b34ba86ed4a38b7d5e0be264207b9451d2/pkg/chartutil/values.go#L281)):*
> is YAML null

This reads strangely. How about "is the YAML null value?"

---

*[pkg/chartutil/values.go, line 286 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2648#-KoNzu04btZFP9iQG8yP:-KoNzu05qmBr62R_pwiO:bd86x9k) ([raw file](https://github.com/kubernetes/helm/blob/413a01b34ba86ed4a38b7d5e0be264207b9451d2/pkg/chartutil/values.go#L286)):*
> ```Go
> 		// ref: http://www.yaml.org/spec/1.2/spec.html#id2803362
> 		switch val {
> 		case "null":
> ```

You can instead write the following:
```go
case "null", "Null", "NULL", "~":
```
Alternately, use [a regular expression](https://play.golang.org/p/VZOYczrtyi).

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2648#-:-KoO6W_mbOg1BpVqZe71:b-o3y4no)*
<!-- Sent from Reviewable.io -->

seh commented:
<img class="emoji" title=":lgtm:" alt=":lgtm:" align="absmiddle" src="https://reviewable.kubernetes.io/lgtm.png" height="20" width="61"/>

---

Reviewed 1 of 1 files at r3.
Review status: all files reviewed at latest revision, 1 unresolved discussion.

---

*[pkg/chartutil/values.go, line 286 at r2](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2648#-KoNzu04btZFP9iQG8yP:-KoOMafNaVghgoJWXyRr:bdfa0rb) ([raw file](https://github.com/kubernetes/helm/blob/413a01b34ba86ed4a38b7d5e0be264207b9451d2/pkg/chartutil/values.go#L286)):*
<details><summary><i>Previously, seh (Steven E. Harris) wrote…</i></summary><blockquote>

You can instead write the following:
```go
case "null", "Null", "NULL", "~":
```
Alternately, use [a regular expression](https://play.golang.org/p/VZOYczrtyi).
</blockquote></details>

I like the regular expression approach because it allows you to write the following:
```go
if stringIsYAMLNull(val) {
  delete(v, key)
  continue
}
```
The single-case `switch` looks a little weird to me, but I concede that the regular expression version also either requires or at least could involve a separate line to compile the regular expression pattern. If you don't want to save the compiled pattern, you can use [`regexp.Match`](https://golang.org/pkg/regexp/#Match) or [`regexp.MatchString`](https://golang.org/pkg/regexp/#MatchString).

---


*Comments from [Reviewable](https://reviewable.kubernetes.io:443/reviews/kubernetes/helm/2648#-:-KoONSam-Q88uRh-99Ks:bnfp4nl)*
<!-- Sent from Reviewable.io -->

scottrigby commented:
@seh I'd be OK with adding the regex check. Honestly that does feel cleaner (and more reusable) to me too. I'll ping you on Slack to make sure I'm 100% clear on how and where you'd like organized in the codebase.

scottrigby commented:
@seh I changed the switch case to a simple in-place version of `regexp.MatchString` rather than adding a new complied `stringIsYAMLNull`. Mainly because I'm not sure whether this has any other use value besides here. Either way, I'd love more feedback on this as this as-is… and then if compiling is preferable I can convert to that.

thomastaylor312 commented on pkg/chartutil/values.go:
```diff
@@ -278,6 +279,18 @@ func coalesceValues(c *chart.Chart, v map[string]interface{}) (map[string]interf
+		// remove incompatible keys from any previous chart, file, or set values.
+		// ref: http://www.yaml.org/spec/1.2/spec.html#id2803362
+		var nullPattern = regexp.MustCompile("^(?:null|Null|NULL|~)$")
+		if str, ok := val.(string); ok {
```
I don't think there is any need for a regex in here. I think the best place to do this would be down on line 297 (in this code) by inserting an `else if ok && v[key] == nil`. The YAML parser already handles all of the null cases and I tested that just to make sure.

scottrigby commented:
@thomastaylor312 OK, I changed to nil comparison. Thanks for checking that this is the case 💯 

If this needs a test, can you point me in the right direction of where to look for the existing tests here? Also, could a test be a follow-up PR, or should it be the same one?

scottrigby commented:
@thomastaylor312 How do these tests look?

thomastaylor312 commented on pkg/chartutil/values_test.go:
```diff
@@ -316,6 +322,8 @@ func TestCoalesceValues(t *testing.T) {
 	}{
 		{"{{.top}}", "yup"},
+		// The keys bottom, right, left, front, and back should not be present.
+		{"{{.side}}", ""},
```
These look great, we do need to test that all of the other keys are unset. This was a learning experience for me too, but I found out it should look like this: `{"{{.bottom}}", "<no value>"},`

scottrigby replied:
I saw that, but doesn't that signify the key is present but the value is empty? Is there currently a test method available that checks if a key is not present?

thomastaylor312 replied:
Definitely a good point. We are talking about this now in Slack to get some ideas

thomastaylor312 commented:
I manually tested this and it is good to go as soon as we have those last unit test cases. Thanks for all the hard work @scottrigby!

scottrigby commented:
@thomastaylor312 OK I updated the PR with the check to ensure keys are not present in the Values map. I just defined an array of the keys that should have null value properly recognized (and converted to `nil` by the yaml parser), and looped over them to check that they are not in the map.

Also, thanks for helping me find what type `Values` is (from `/pkg/chartutil/values.go`: `type Values map[string]interface{}`. From the documentation of `CoalesceValues` it was not clear to me that `Values` referred to a type declaration).

scottrigby committed:
New Helm functionality to delete a coalesced YAML key when the value is nil.

- Note that this covers all YAML null syntax options:
  ref: http://yaml.org/type/null.html
- Note that we do a nil comparison because the encoding/yaml package parses
  YAML properly and any variation of null, Null, NULL, or ~ is converted to nil
  by the time we get here.

scottrigby committed:
Test that the coalesced key is properly removed for all YAML null syntax options, but not for empty string

scottrigby committed:
Update documentation for deleting a default key from a values file

scottrigby commented:
Note that the documentation is the only clear example I can think of now, because it's from a PR I have worked on https://github.com/kubernetes/charts/pull/1375, but that PR is not yet merged. However, it clearly explains the functionality, and I tried to make it match the rhetorical style of the rest of the doc (except I added a header for clarity). If there is another, better example, we can use that. But it does feel like an example helps. Thoughts?

thomastaylor312 commented:
The docs look good to me. Because circle isn't running the unit tests, I am running those locally and validating that they pass

scottrigby commented:
Rebased again against `kubernetes/master` now.

thomastaylor312 commented:
Ok, I checked that the tests are passing, so this should be good to go. Thanks for the contribution @scottrigby!

seh commented on pkg/chartutil/values.go:
```diff
@@ -281,6 +281,13 @@ func coalesceValues(c *chart.Chart, v map[string]interface{}) (map[string]interf
 		if _, ok := v[key]; !ok {
 			// If the key is not in v, copy it from nv.
 			v[key] = val
+		} else if ok && v[key] == nil {
```
- [ ] Here, we could have avoided a second map lookup by binding the mapped value to a variable at line 281 above:
```go
if value, ok := v[key]; ! ok {
  // ...
} else if ok && value == nil {
  // When the YAML ...
}
```
I'd consider then turning the conditional around:
```go
if value, ok := v[key]; ok {
  if value == nil {
    // When the YAML ...
  } else if dest, ok := value.(map[string]interface{}); ok {
    // If v[key] is a table ...
  }
} else {
  // If the key is not in v ...
}
```

thomastaylor312 replied:
@seh If you create an issue and assign it to me, I can make these small changes

seh replied:
That became #2663.

seh commented on pkg/chartutil/values.go:
```diff
@@ -281,6 +281,13 @@ func coalesceValues(c *chart.Chart, v map[string]interface{}) (map[string]interf
+			// When the YAML value is null, we remove the value's key.
+			// This allows Helm's various sources of values (value files or --set) to
+			// remove incompatible keys from any previous chart, file, or set values.
+			// ref: http://www.yaml.org/spec/1.2/spec.html#id2803362
```
- [ ] I'd prefer to move this reference over into the pertinent test function in file _values_test.go_ (around line 280 there).

seh reviewed (commented):
I missed the merge here, but thought it was still worth noting these suggestions.

seh commented:
By the way, the `go-yaml/yaml` package uses [a map of strings to resolve special identifiers to YAML tags](https://github.com/go-yaml/yaml/blob/4c78c975fe7c825c6d1466c42be594d1d6f3aba6/resolve.go#L18). [Here](https://github.com/go-yaml/yaml/blob/4c78c975fe7c825c6d1466c42be594d1d6f3aba6/resolve.go#L43) we find the set that maps to the "null" tag.
